dashboard/test/test_helper.rb

```diff
@@ -249,22 +249,78 @@ def assert_redirected_to_sign_in
   end
 
   def self.generate_admin_only_tests_for(action, params = {})
-    test "should get #{action}" do
-      sign_in create(:admin)
-      get action, params: params
-      assert_response :success
+    test_user_gets_response_for action, user: :admin, params: params
+    test_user_gets_response_for action, response: :forbidden, params: params
+    test_redirect_to_sign_in_for action, params: params
+  end
+
+  # Generates a test case ensuring redirect to sign in for not signed in users
+  # @param action [String] the controller action to test
+  # @param method [Symbol, String] http method with which to perform the action (default :get)
+  # @param params [Hash, Proc] params to pass to the action. It can be a direct hash,
+  #   or a proc that generates a hash at runtime in the context of the test case.
+  def self.test_redirect_to_sign_in_for(action, method: :get, params: {})
+    test_user_gets_response_for(
+      action,
+      name: "not signed in #{method} #{action} redirects to sign in",
+      method: method,
+      params: params,
+      response: :redirect
+    ) do
+      assert_redirected_to_sign_in
     end
+  end
 
-    test "should not get #{action} if not signed in" do
-      sign_out :user
-      get action, params: params
-      assert_redirected_to_sign_in
+  # Generates a basic response validation test case for a user, logged-in or not.
+  # @param action [String] the controller action to test
+  # @param method [Symbol, String] http method with which to perform the action (default :get)
+  # @param response [Symbol, String, Number] expected response (default :success)
+  # @param user [Symbol, String, Proc] user to log in, or nil to test as a not-logged-in user (default: nil)
+  #   It can be a factory name to be passed to FactoryGirl.create,
+  #   or a proc that runs in the context of the test case and returns a user.
+  # @param params [Hash, Proc] params to pass to the action. It can be a direct hash,
+  #   or a proc that generates a hash at runtime in the context of the test case.
+  # @param name [String] optional name of the test case.
+  #   Otherwise it will be generated based on parameter values.
+  #   Note: It is recommended to supply a name when varying params or user procs,
+  #     as the default generated names may be ambiguous and may conflict.
+  # @yield runs an optional block of additional test logic after asserting the response
+  #   Note: name is required when providing a block.
+  def self.test_user_gets_response_for(action, method: :get, response: :success,
+    user: nil, params: {}, name: nil, &block)
+
+    unless name.present?
+      raise 'name is required when a block is provided' if block
+      user_display_name =
+        if user.is_a?(Proc)
+          'supplied user'
+        elsif user.present?
+          user
+        else
+          'not logged-in user'
+        end
+
+      name = "#{user_display_name} calling #{method} #{action} should receive #{response}"
```

Comment:
It seems like there is a small risk of name collisions here if you change the parameters but not the user, method, action, or response. I'm not sure it makes sense to put the params in the test name so I think this is probably ok.